### ISSUE TITLE
 3.0.0 release not Strong Name signed?

### ISSUE THREAD

**Issue by kgoderis** [Azure/azure-webjobs-sdk-extensions#504]
I am definitely not an expert in Strong Naming DLL's but I somehow noticed that the 3.0.0 DLLs are only delay signed, and not strong named. As we have strong named code that is making calls to Microsoft.Azure.WebJobs.Extensions.Http for example, our code will obviously not run anymore. 

Falling back to 3.0.0-rc1 fixes the problem for now. 

Was the strong naming perhaps accidentally forgotten?

**Comment by paulbatum** [comment]
Can you include more details on why you believe this is the case? Here are the verification steps I've taken:

1. Downloaded the nuget package, ran `sn -vf` on the binary:

```
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise>sn -vf D:\downloads\microsoft.azure.webjobs.extensions.http.3.0.0\lib\netstandard2.0\Microsoft.Azure.WebJobs.Extensions.Http.dll

Microsoft (R) .NET Framework Strong Name Utility  Version 4.0.30319.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Assembly 'D:\downloads\microsoft.azure.webjobs.extensions.http.3.0.0\lib\netstandard2.0\Microsoft.Azure.WebJobs.Extensions.Http.dll' is valid
```

2. Made a brand new V2 function app in Visual Studio, go into the project and enable signing, build, run `sn -vf` on the output:

![image](https://user-images.githubusercontent.com/10663/46239113-c1209600-c348-11e8-88f3-4a60d1abdadb.png)

```
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise>sn -vf C:\Users\pbatum\source\repos\FunctionApp65\FunctionApp65\bin\Debug\netstandard2.0\bin\FunctionApp65.dll

Microsoft (R) .NET Framework Strong Name Utility  Version 4.0.30319.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Assembly 'C:\Users\pbatum\source\repos\FunctionApp65\FunctionApp65\bin\Debug\netstandard2.0\bin\FunctionApp65.dll' is valid
```

3. Modified my function app project to delay signed, to confirm that `sn -vf` detects this and reports it appropriately:

```
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise>sn -vf C:\Users\pbatum\source\repos\FunctionApp65\FunctionApp65\bin\Debug\netstandard2.0\bin\FunctionApp65.dll

Microsoft (R) .NET Framework Strong Name Utility  Version 4.0.30319.0
Copyright (c) Microsoft Corporation.  All rights reserved.

C:\Users\pbatum\source\repos\FunctionApp65\FunctionApp65\bin\Debug\netstandard2.0\bin\FunctionApp65.dll is a delay-signed or test-signed assembly
```

Do you get different results when you perform these steps on your machine? Any other steps you can suggest I try?

**Comment by kgoderis** [comment]
@paulbatum I not familiar with MS's tooling - I use VS Code primarily.

Azure Functions Core Tools (2.0.2.0)
Function Runtime Version: 2.0.12115.0

1. In .csproj I have:

    <PackageReference Include="Microsoft.Azure.WebJobs.Extensions.Http" Version="3.0.0" />

2. I initiate a Build Task, e.g. "dotnet clean" and "dotnet build", and after a successful build I have in my logs  SNDOdoo -> /Users/kgoderis/SNDOdoo/bin/Debug/netstandard2.0/bin/SNDOdoo.dll

3. In a Terminal, I perform

C02STBWMGTFM:bin kgoderis$ cd  /Users/kgoderis/SNDOdoo/bin/Debug/netstandard2.0/bin
C02STBWMGTFM:bin kgoderis$ sn -v Microsoft.Azure.WebJobs.Extensions.Http.dll
Mono StrongName - version 4.8.0.0
StrongName utility for signing assemblies
Copyright 2002, 2003 Motus Technologies. Copyright 2004-2008 Novell. BSD licensed.

Assembly Microsoft.Azure.WebJobs.Extensions.Http.dll is delay-signed but not strongnamed

The final binary (SNDOdoo.dll in this case) is strong named when I check it with sn -v

Now, I just noticed that I get the same result when using 3.0.0-rc1, but the main difference is that the unsigned code that is called, *is* working in this case. I don't understand what makes the difference here.

**Comment by paulbatum** [comment]
So I take it you're running on a mac, given that it looks like a mono version of the strong name utility was used.

Lets move away from the output of sn.exe for a moment and focus on your actual problem. You said your code isnt running, but you never included the error message. Could you try to find a set of standalone repro steps that we can try on a mac, similar to what I did above in step 2.

**Comment by kgoderis** [comment]
@paulbatum I will try to do that, but to give you some more context. 

I make calls to an external (commercial) library (https://www.nuget.org/packages/WW.Cad.NetStandard) for which do not have access to the source code. The library works by means of a license key which is matched against a strong name key that the purchasing party has to use to to sign its own DLL (in my case, using the `[assembly: AssemblyKeyFileAttribute(keyFile: "SNDOdoo.snk")]` attribute). This verification step is done by a method called SetLicense(). As far as I know, when using a "trial" license for this library the setLicense code opens a browser window on the calling host, and redirects to a webpage of the developer, in order for the calling party to obtain a full license key. I.e. it most likely uses Extensions.Http to make that call. 

When I run my binary using 3.0.0, then the SetLicense methods simply hangs (no exception or stack trace thrown). When running against 3.0.0-rc1, the call to SetLicense succeeds without a problem. Since I do not have the source of the library, I have no clue what is really making a difference here.

**Comment by paulbatum** [comment]
OK, I followed most of that, except for this part:
>  when using a "trial" license for this library the setLicense code opens a browser window on the calling host, and redirects to a webpage of the developer, in order for the calling party to obtain a full license key. I.e. it most likely uses Extensions.Http to make that call.

The azure webjobs http extension has nothing to do with opening a browser.

One thing I'll note about the http extension is that its a bit unusual in that its actually built in to functions itself. You reference it in your code so that you can use the types it provides (such as `HttpTriggerAttribute`) but at runtime, the version that is loaded is the version that is built in. Given that you are running `2.0.12115.0`, this means that the GA version of extensions.http should be loaded and not the RC version that you're referencing. I don't know how this explains the behavior that you are seeing.

One other possibility occurs to me - if you ever referenced any of our internal nuget feeds, you might have downloaded and cached a non-signed version. Either of these feeds might have caused that to happen:
https://www.myget.org/F/azure-appservice/api/v3/index.json  
https://www.myget.org/F/azure-appservice-staging/api/v3/index.json  

Maybe clear out your local nuget cache? Another way to verify would be to do my step 1 above (download directly from nuget using the browser, unzip the nuget, run sn on the unzipped binary).

**Comment by kgoderis** [comment]
@paulbatum I dowloaded the nupkg manually, then added it to a local Source, then made that source the first entry in the NuGet.Config file, cleaned out the local cache, and did a restore.

After compiling the project I still get:

```
C02STBWMGTFM:bin kgoderis$ versioninfo  Microsoft.Azure.WebJobs.Extensions.Http.dll
/Users/kgoderis/SNDOdoo/bin/Debug/netstandard2.0/bin/Microsoft.Azure.WebJobs.Extensions.Http.dll
	FileVersionInfo.FileVersion:	3.0.0.0
	FileVersionInfo.ProductVersion:	3.0.0 Commit hash: f45a148e9f2ada2f4dcd977aa59b50d254f1a72a
C02STBWMGTFM:bin kgoderis$ sn -vf  Microsoft.Azure.WebJobs.Extensions.Http.dll
Mono StrongName - version 4.8.0.0
StrongName utility for signing assemblies
Copyright 2002, 2003 Motus Technologies. Copyright 2004-2008 Novell. BSD licensed.

Assembly Microsoft.Azure.WebJobs.Extensions.Http.dll is delay-signed but not strongnamed
```

**Comment by paulbatum** [comment]
I think there must be a bug with the mono version of sn.exe. I just tried it myself and I got the same message as you, but then I ran it on a webjobs 2.x assembly and again I got the same output. The webjobs 2.x assemblies have been strong named for a long time now and we know there are no issues with it because customers have been strong naming their own applications without issue.

Here is one more test I did. I made a .NET framework 4.6.1 app and had it reference the older beta8 nuget package for the http extension. I configured my console app to be strong name signed, and I got the following build warning and the following runtime failure:

```
BUILD TIME:
Referenced assembly 'Microsoft.Azure.WebJobs.Extensions.Http, Version=3.0.0.0, Culture=neutral, PublicKeyToken=null' does not have a strong name.	

RUN TIME:
System.IO.FileLoadException
  HResult=0x80131044
  Message=Could not load file or assembly 'Microsoft.Azure.WebJobs.Extensions.Http, Version=3.0.0.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies. A strongly-named assembly is required. 
```

I then updated my package reference to the GA package, rebuilt and ran and both the build warning and the runtime error disappeared. This confirms that the .NET framework has no issue with strong name verification of the assembly.

I'm inclined to say that the output from the mono version of sn is a red herring, and perhaps the hang you are getting is related to some other issue...
